application/src/main/java/org/togetherjava/tjbot/commands/utils/StringDistances.java

```diff
@@ -51,6 +51,29 @@ public static <S extends CharSequence> Optional<S> autocomplete(CharSequence pre
             .min(Comparator.comparingInt(candidate -> prefixEditDistance(prefix, candidate)));
     }
 
+    /**
+     * Gives sorted suggestion to autocomplete a prefix string from the given options
+     *
+     * @param prefix the prefix to give suggestion for
+     * @param candidates all the possible suggestions
+     * @param errorMargin error margin of suggestions, error percentage / 100
+     * @return collection of autocomplete suggestions
+     */
+    public static Collection<String> autocompleteSuggestions(CharSequence prefix,
+            Collection<String> candidates, double errorMargin) {
+        int firstCandidateScore = candidates.stream()
+            .mapToInt(candidate -> prefixEditDistance(prefix, candidate))
+            .min()
+            .orElse(0);
+        return candidates.stream()
+            .map(candidate -> new CandidateScore(candidate, prefixEditDistance(prefix, candidate)))
+            .sorted()
+            .takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore
+                    + firstCandidateScore * errorMargin)
+            .map(CandidateScore::candidate)
+            .toList();
+    }
```

Comment:
the whole design for this method is not really "idiomatic". and there are multiple issues with it. lets refactor it a bit:
```java
public static Collection<String> closestMatches(CharSequence source, Collection<String> destinations, int limit) {
        // TODO some argument validation ...
        
        Queue<String> matches = new PriorityQueue<>(Comparator.comparingInt(x -> prefixEditDistance(source, x)));
        matches.putAll(destinations);
        
        return Stream.generate(matches::poll)
          .limit(limit)
          .toList();
}
```

Reply:
Can you elaborate on the issues, as it seems youre removing part of the functionality.

There's a reason it gets mapped to a CandidateScore, take a look at the takeWhile method.

Reply:
The way this math attempts to provide a good UX isnt really working out imo. Its completely missing the string length for example.

I am okay with adding this feature again. But then incorperate it into the proposed design with the PQ instead. The `CandidateScore` class should be nested, nobody else needs it. And should be renamed to `MatchScore` or `ScoredMatch`, depending on what it contains.

Reply:
I certainly believe this is far from the best solution.
but I do believe it's better than your suggestion of deleting it completely.

Since I personally lack experience with algorithms, you're on your own here.

Reply:
To begin with, i dont even understand how this math is supposed work at all. Its definitely in need for its own little method/comments/javadoc. And I would appreciate it if someone can give me a quick example calculation, so that i understand it.

Reply:
probably bc u made it public? hard to tell for me without seeing what u saw.

Reply:
should i put the record in the method?

Reply:
not resolved. still missing the priority queue approach with k-sort. and all the explanation/rework regarding that odd math.

Reply:
I thought u and tais agreed on keeping the current implementation, sry english is not my first language

Reply:
no worries. we agreed on keeping a filtering to not display non-sense suggestions. but:
* this can still be added to the PQ k-sort approach perfectly fine
* and im not okay with this odd way of doing it and would like to get a small explanation of how it works to understand it and then suggest something better

Reply:
The idea of this was made up by me, I don't know or there's any official algorithm like this.

The most confusing method is 
```java
            .takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore
                    + firstCandidateScore * errorMargin)
```

The idea of this method is pretty simple.

If the most accurate completion is 2 characters away, the most inaccurate completion can only be a max of 6 characters away. A slight improvement would do this.

```java
            .takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore * 2 * errorMargin)
```

Maybe a helper method would help even more?

```java
isCandidateInErrorRange(candidateScore , firstCandidateScore, errorMargin)
```

Do you understand it now?

Reply:
sorry, no. can u give me a concrete example with some toy data?

Reply:
We have input `autocomplet`

the most accurate completion is `autocomplete` which is a total of 1 character away

1 * 2 * 0.5 =  3
The farthest distance a completion can be away is 3, so any completion that has a stringdistance of `1/2/3` will be accepted.



Another completion, `autocom` is total of 4 character, this means this won't get accepted.
But `autocomp` will get accepted, since it's only 3 characters away.


This way, we won't just have 25 random completion.
Makes more sense?

Reply:
yes. i understand the math now. it feels reversed to what i would have done:
```java
top10Candidates
  .filter(candidateWithPed -> candidateWithPed.ped / input.length()  < 0.5)
```
example, user inputs `"aaa"` and we have candidates `"aab"` and `"foo"`. latter is totally off, the PEDs are:
* `"aab" -> 1
* `"foo"` -> 3

the filter divides the PED with the original length `"aaa".length() -> 3` to compute how far away the candidate is:
* `"aab" -> `1 / 3` -> `0.333` is `< 0.5`, accepted
* `"foo"` -> `3 / 3` -> `1.0` is not `< 0.5`, discarded

Reply:
I'm noticing some knowledge about algorithms, sounds better than my suggestion. \:p

Reply:
this can ofc be optimized to already execute on the original stream with a `takeWhile`, like u suggested.

Reply:
```java
.takeWhile(candidateScore -> candidateScore.score() <= firstCandidateScore * 2 * errorMargin)
```
@Tais993 this is not the same thing

Reply:
what's the difference between `firstCandidateScore  * 2` and `firstCandidateScore + firstCandidateScore `

Reply:
Actually, the way it was written (urs, but also mine) both have a performance flaw. We have to compute the PED for all entries once and then remember that instead. Using the PED in the Comparator will result in it being recomputed many times (more for sorted() than for a PQ, but still way too often). So we should do it like this:
```java
List<MatchScore> scoredMatches = getAllTags().stream()
  .map(match -> computePed(input, match))
  .toList();
```
and now we can sort based on the already-computed score:
```java
Queue<MatchScore> bestMatches = new PriorityQueue<>();
bestMatches.addAll(scoredMatches);

List<String> results = Stream.generate(bestMatches::poll)
  .takeWhile(match -> match.score / input.length() <= 0.5)
  .map(MatchScore::getMatch)
  .limit(10)
  .toList();
```
or similar (writing this from memory)

Reply:
> what's the difference between `firstCandidateScore * 2` and `firstCandidateScore + firstCandidateScore `

thats not what it is, i first multiply errorMargin in firstCandidateScore and then add it to firstCandidateScore

Reply:
Ah I completely forgot about mathematical rules LMFAO

Reply:
Why not just do `firstCandidateScore * (errorMargin + 1)`, since this is identical